### PR TITLE
[6X] Retry gang creation for non-recovery failures

### PR DESCRIPTION
**Pull request by soumyadeep2007**
This is a backport of 7X commit:
99a48e65b803ae7e446894177f4f97c10415ac18 with one very notable change:
the GUC gp_gang_creation_retry_non_recovery controls whether we want to
retry on non-recovery failures, giving users a chance to opt out of the
behavior.

Original commit message follows:

Before this commit, any error from a segment that is unrelated to
recovery, would result in an immediate termination with the following:

ereport(ERROR, (errcode(ERRCODE_GP_INTERCONNECTION_ERROR),
	errmsg("failed to acquire resources on one or more segments"),
	errdetail("%s (%s)", PQerrorMessage(segdbDesc->conn), segdbDesc->whoami)));

In other words, this termination did not respect
gp_gang_creation_retry_count. From now on, we exhaust the retries first,
before emitting this ERROR.

Co-authored-by: Ashwin Agrawal <aashwin@vmware.com>
Reviewed-by: Xin Zhang <xinzweb@users.noreply.github.com>
Reviewed-by: Huansong Fu <fuhuansong@gmail.com>

### CODE REVIEW

**Comment by soumyadeep2007** [comment]
The failure flagged by the ABI checker is a false positive:
https://github.com/greenplum-db/gpdb/actions/runs/6881339997/job/18717810377?pr=16741#step:8:74

We can always add a new GUC and the struct array size will change. Unavoidable.

**Review by thovhannes — APPROVED** [review]

File: src/backend/utils/misc/guc_gp.c [hunk 1]
```diff
@@ -913,6 +913,16 @@ struct config_bool ConfigureNamesBool_gp[] =
...
 	},
 
+	{
+		{"gp_gang_creation_retry_non_recovery", PGC_USERSET, QUERY_TUNING_METHOD,
+		 gettext_noop("Retry gang creation if non-recovery failures are encountered during dispatch."),
+		 NULL,
+		 GUC_NO_SHOW_ALL | GUC_NOT_IN_SAMPLE
+		},
+		&gp_gang_creation_retry_non_recovery,
+		true,
```
Comment (huansong) [inline src/backend/utils/misc/guc_gp.c:923]: I thought we were to keep 6X behavior same as before, and let user opt-in to turn on this feature? If so shouldn't the GUC default to be false?

Also, in the current change what would happen when there's a non-recovery fault but `gp_gang_creation_retry_non_recovery` is false? It seems that we would fail the `Assert(successful_connections + in_recovery_mode_count + other_failures == size)` because we haven't incremented `other_failures`?

**Review by huansong — COMMENTED** [review]
Some questions about the GUC default value. Otherwise LGTM.

> We can always add a new GUC and the struct array size will change. Unavoidable.

Agreed. We should probably have the ABI check ignore these variables somehow. Otherwise would be pretty annoying going forward.

File: src/backend/utils/misc/guc_gp.c (hunk 1 above)
Comment (soumyadeep2007) [reply src/backend/utils/misc/guc_gp.c:923]: > I thought we were to keep 6X behavior same as before, and let user opt-in to turn on this feature? If so shouldn't the GUC default to be false?

Actually, we want this on by default even on 6X. As discussed in the internal meeting, since this was reported on 6X, we want this to be on by default. This on behavior will be the more desirable one and in the edge case it isn't, we have provided a mechanism to opt out.

> Also, in the current change what would happen when there's a non-recovery fault but gp_gang_creation_retry_non_recovery is false? It seems that we would fail the Assert(successful_connections + in_recovery_mode_count + other_failures == size) because we haven't incremented other_failures?

Good catch! I missed adding a hunk. Fixed.

**Comment by soumyadeep2007** [comment]
> Agreed. We should probably have the ABI check ignore these variables somehow. Otherwise would be pretty annoying going forward.

Let me reach out to the team if they know a way to do that.
